src/main/java/life/qbic/model/download/QbicDataDownloader.java

```diff
@@ -227,6 +227,35 @@ private void downloadFileFromDataStoreServer(DataSetFile dataSetFile, Path prefi
         final Path finalPath = OutputPathFinder.determineOutputDirectory(outputPath, prefix,
             file.getDataSetFile(), conservePaths);
         File newFile = new File(finalPath.toString());
+
+        if (newFile.exists()) {
+          try (
+              CheckedInputStream existingFileReader = new CheckedInputStream(
+                  Files.newInputStream(newFile.toPath()), new CRC32())) {
+
+            int bufferSize =
+                (newFile.length() < defaultBufferSize)
+                    ? (int) file.getDataSetFile().getFileLength()
+                    : defaultBufferSize;
+            byte[] buffer = new byte[bufferSize];
+            while (existingFileReader.read(buffer) != -1) {
+              // reading
+            }
+
+            ChecksumValidationResult checksumValidationResult = validateChecksum(
+                existingFileReader.getChecksum().getValue(), dataSetFile);
+            if (checksumValidationResult.isValid()) {
+              LOG.info("Found existing file with the identical content. Skipping "
+                  + finalPath.toAbsolutePath());
+              return;
+            } else {
+              LOG.info("Updating existing file " + finalPath.toAbsolutePath());
+            }
+          } catch (IOException e) {
+            throw new RuntimeException("Existing file could not be processed", e);
+          }
+        }
```

Comment:
I think the method can be refactored to increase readability (including strategy pattern or pack the logic in separate methods).

I also see a violation in the SRP (not even, it is on function level: a function should do only one thing): the function does download, checksum verification and update file decision. These are three responsibilities :D Maybe you find a strategy to resolve this on the application level, not on the I/O level?

Reply:
Resolved in personal discussions: we accept the change, since postman need to be refactored for the new portal anyway.